### PR TITLE
Iterative beam search in TorchGeneratorAgent

### PR DESCRIPTION
**Pull request by uralik**
With added arguments `beam-num-iterations`, `beam-min-distance` user can turn it on. This now only supports batch size 1. It can be done with mini-batches, it can be extended for that later.

Multiple iterations are implemented as a mini-batch beam, so very little change in the agent code as well as in the Beam class code.

Right now it just returns the best candidate given the logprob score, but intended to be changed by the user in any way (with external re-ranker, any other piece of code).

### CODE REVIEW

**Comment by uralik** [comment]
lint is fine locally

**Comment by emilydinan** [comment]
@uralik if you click on the "Details" in the travis build failure you can see the lint errors it's complaining about:

```
parlai/core/torch_generator_agent.py:862:21: F841 local variable 'current_hyp_length' is assigned to but never used
parlai/core/torch_generator_agent.py:865:89: E501 line too long (98 > 88 characters)
```

File: parlai/core/torch_generator_agent.py [hunk 1]
```diff
@@ -773,13 +802,22 @@ def __init__(self, beam_size, min_length=3, padding_token=0, bos_token=1,
...
+        self.min_distance = min_distance
 
     @staticmethod
     def find_ngrams(input_list, n):
         """Get list of ngrams with context length n-1"""
         return list(zip(*[input_list[i:] for i in range(n)]))
 
+    def hamming_dist(self, l1, l2):
+        assert len(l1) == len(l2), 'Hamming distance requires same length lists'
+        res = 0
```
Comment (emilydinan) [inline parlai/core/torch_generator_agent.py]: can do this in one line with `sum(a != b for a, b in zip(l1, l2))`

**Comment by uralik** [comment]
@emilydinan yeah I know :) thats why I wrote that local execution of lint script doesnt give errors, will fix those you mentioned.

File: parlai/agents/seq2seq/modules.py [hunk 2]
```diff
@@ -453,7 +453,7 @@ def forward(self, input):
             scores = F.linear(e, self.weight, self.bias)
 
         if self.padding_idx >= 0:
-            scores[:, :, self.padding_idx] = -NEAR_INF
+            scores[:, self.padding_idx] = -NEAR_INF
```
Comment (emilydinan) [inline parlai/agents/seq2seq/modules.py]: was this broken before?

Comment (emilydinan) [reply parlai/agents/seq2seq/modules.py]: cc @stephenroller @alexholdenmiller

Comment (uralik) [reply parlai/agents/seq2seq/modules.py]: seems like that. This thing is used only when lookup table is shared, so prolly was missed accidentally..

Comment (emilydinan) [reply parlai/agents/seq2seq/modules.py]: 😢  thanks for fixing!

Comment (stephenroller) [reply parlai/agents/seq2seq/modules.py]: Are we sure about that? scores should be a (bsz x seqlen x vocab) tensor.

Comment (stephenroller) [reply parlai/agents/seq2seq/modules.py]: Could be an issue because you limited to bsz 1? Is there an unfortunate squeeze somewhere?

File: parlai/core/torch_generator_agent.py [hunk 3]
```diff
@@ -624,7 +647,12 @@ def beam_search(self, model, batch, beam_size, start=1, end=2,
             score = F.log_softmax(score, dim=-1)
             for i, b in enumerate(beams):
                 if not b.done():
-                    b.advance(score[i])
+                    partial_hyps = []
+                    if num_iterations > 1:
+                        for j in range(i):
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: `for j in (index into beams)` seems strange. Is this intentional?

File: parlai/core/torch_generator_agent.py [hunk 4]
```diff
@@ -788,7 +826,7 @@ def get_backtrack_from_current_step(self):
         """Get the backtrack at the current step."""
         return self.bookkeep[-1]
 
-    def advance(self, softmax_probs):
+    def advance(self, softmax_probs, neighbor_partial_hyps):
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: Add a brief docstring please.

File: parlai/agents/seq2seq/modules.py (hunk 2 above)
Comment (uralik) [reply parlai/agents/seq2seq/modules.py]: discussed offline with @stephenroller , likely happens due to bsz 1 and singleton dimension got squeezed.

File: parlai/core/torch_generator_agent.py [hunk 5]
```diff
@@ -807,7 +845,7 @@ def advance(self, softmax_probs):
                            self.scores.unsqueeze(1).expand_as(softmax_probs))
             for i in range(self.outputs[-1].size(0)):
                 if self.block_ngram > 0:
-                    current_hypo = self.partial_hyps[i][1:]
+                    current_hypo = self.partial_hyps[i]
```
[the remark this answers is not in the document]
Comment (uralik) [reply parlai/core/torch_generator_agent.py]: There is a change in the init() : self.partials does not have bos there, so we dont need this bos slicing out.

File: parlai/core/torch_generator_agent.py [hunk 6]
```diff
@@ -515,11 +524,19 @@ def eval_step(self, batch):
...
+                min_distance=self.beam_min_distance
             )
             beam_preds_scores, _, beams = out
             preds, scores = zip(*beam_preds_scores)
 
+            if self.beam_num_iterations > 1:
+                # here the best candidate w.r.t. logprob is selected
+                # user can implement any other way to choose the best one
+                best_score_idx = torch.argmax(torch.Tensor(scores))
+                preds, scores = [preds[best_score_idx]], [scores[best_score_idx]]
```
[the remark this answers is not in the document]
Comment (uralik) [reply parlai/core/torch_generator_agent.py]: your line will not work since later it expects preds to be a *list* of predictions, thats why I put that list wrapping line 538. But thats also the effect of bsz 1 :)

File: parlai/core/torch_generator_agent.py (hunk 3 above)
Comment (uralik) [reply parlai/core/torch_generator_agent.py]: so yes we iterate over all iterations done BEFORE i

File: parlai/core/torch_generator_agent.py [hunk 7]
```diff
@@ -617,14 +642,19 @@ def beam_search(self, model, batch, beam_size, start=1, end=2,
 
             score, incr_state = model.decoder(decoder_input, encoder_states, incr_state)
             # only need the final hidden state to make the word prediction
-            score = score[:, -1, :]
+            score = score[:, -1, :].unsqueeze(1)
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: This also seems surprising to me, that the bug is here. I'll need to investigate further.

File: parlai/core/torch_generator_agent.py [hunk 8]
```diff
@@ -773,13 +804,22 @@ def __init__(self, beam_size, min_length=3, padding_token=0, bos_token=1,
         self.n_best_counter = 0
         self.min_n_best = min_n_best
         self.block_ngram = block_ngram
-        self.partial_hyps = [[self.bos] for i in range(beam_size)]
+        self.partial_hyps = [[] for i in range(beam_size)]
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: Does bos still get inserted as an input? I think seq2seq and transformer/generator have different behaviors w.r.t. this.

File: parlai/core/torch_generator_agent.py (hunk 7 above)
Comment (uralik) [reply parlai/core/torch_generator_agent.py]: steps to get this thing happened:

1. train the model like this (should be newlines there...):
training:
`python train.py \
    -dt train \
    -t convai2:self \
    -m seq2seq \
    -bs 128 \
    -dp "${DATA_PATH}" \
    --model-file "${RUN_DIR}"/model \
    --dict-file "${DATA_PATH}"/convai2.dict \
    --numlayers 2 \
    --hiddensize 512 \
    --dropout 0.5 \
    --optimizer adam \
    -lr 0.001 \
    -esz 300 \
    --attention general \
    -lt enc_dec \
    -emb glove \
    -veps 5`

2. test the model:
`python eval_model.py \
    -dt valid \
    -t convai2:self \
    -m seq2seq \
    -bs 128 \
    -dp "${DATA_PATH}" \
    --model-file PATH \
    --dict-file "${DATA_PATH}"/convai2.dict \
    --beam-size 2 \`

**Comment by stephenroller** [comment]
Abandoning
